src/core/lib/gprpp/directory.h

```diff
@@ -0,0 +1,46 @@
+//
+//
+// Copyright 2023 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//
+
+#ifndef GRPC_SRC_CORE_LIB_GPRPP_DIRECTORY_H
+#define GRPC_SRC_CORE_LIB_GPRPP_DIRECTORY_H
+
+#include <grpc/support/port_platform.h>
+
+#include <string>
+#include <vector>
+
+#include "absl/status/statusor.h"
+#include "absl/strings/string_view.h"
+
+namespace grpc_core {
+
+class Directory {
+ public:
+  explicit Directory(absl::string_view directory_path)
+      : directory_path_(directory_path) {}
+  virtual ~Directory() = default;
+  virtual absl::StatusOr<std::vector<std::string>> GetFilesInDirectory();
```

Comment:
Do you think it makes more sense for a `DirectoryReader` to return the list of files or to invoke callbacks for each file? I agree with the unnecessary allocation, but it does put more constraints on the code structure of the caller to have to do it in a callback fashion

Reply:
What constraints does it place on the caller?  Instead of doing this:

```
auto results = directory_reader->GetFilesInDirectory();
if (!results.ok()) return results.status();
for (const auto& filename : *results) {
  // ...process file...
}
```

The caller can instead say:

```
absl::Status status = directory_reader->ForEach(
    [&](absl::string_view filename) {
      // ...process file...
    });
if (!status.ok()) return status;
```

Reply:
So I think it's nearly identical in this use case, but if it's more generic I could see a case where someone would want to post-process directory names in a way that doesn't necessarily work well with a `ForEach`. But, then again, then they could just have the callback fill a vector.

The other consideration goes to your comments earlier on keeping this interface very simple and not trying to do anything too smart or clever - the simplest thing for a `DirectoryReader` to do is read a directory, not provide a callback interface to act on each entry? WDYT?

I see value in the callback too though, either is good with me. Or would you want to keep both?

Reply:
> So I think it's nearly identical in this use case, but if it's more generic I could see a case where someone would want to post-process directory names in a way that doesn't necessarily work well with a `ForEach`. But, then again, then they could just have the callback fill a vector.

The only kind of post-processing I can think of that couldn't be done in the lambda is if the caller needs to sort the entries lexicographically.  That use-case seems unlikely in gRPC, but as you pointed out, if we did need that, we could easily do it by having the lambda add each entry to a vector.

> The other consideration goes to your comments earlier on keeping this interface very simple and not trying to do anything too smart or clever - the simplest thing for a `DirectoryReader` to do is read a directory, not provide a callback interface to act on each entry? WDYT?

My argument about not trying to be clever was about functionality, not API design.  I was saying that we shouldn't do unnecessary checks about the state of the filesystem, since those checks would have actually made the functionality worse without adding any benefit.  The reason they didn't add benefit is that they aren't guaranteed to catch all problems with the operations that we actually need to do (e.g., if the parent directory of the directory we're reading is readable but not accessible, the check that the directory exists would have succeeded, but we still wouldn't be able to read the directory).  And making too many assumptions about the state of the filesystem can actually cause us to fail in cases that would not actually cause problems for the operations that we do actually care about (e.g., the way that the `S_ISREGFILE()` check would have skipped symlinks).  If we want things to work in the maximum number of cases with the minimal amount of integration problems, then keeping it simple is important.

In contrast, the question here is about API design, not functionality.  I don't think the argument against those filesystem checks is an argument for any particular API here; no matter what API we choose, the functionality will be the same.

Also, it seems like you might be assuming that the lambda approach is somehow harder or more complex, and I really don't think it is; it's just another way of structuring the interface.  I don't think it adds more code to either the caller or the implementation, and it significantly reduces memory usage.

> I see value in the callback too though, either is good with me. Or would you want to keep both?

I think all we need is the `ForEach()` interface; I don't think we need to retain the `GetDirectoryContents()` interface.

Note that if at some point in the future, we discover that we have a whole bunch of callers of this API that are actually having the lambda dump the entry into a vector, we will be able to trivially add a new non-virtual method on the base class that will implement the `GetDirectoryContents()` semantics as a simple wrapper around `ForEach()`, and it will not be a breaking change.  In contrast, if the only interface that we provide here is `GetDirectoryContents()` and then we discover a case where the extra memory usage is a real problem, then it will be a lot harder to fix it, because you can't implement `ForEach()` in terms of `GetDirectoryContents()`.

Reply:
That all makes sense to me - made the change